### PR TITLE
Never generate lambas, only function definitions

### PR DESCRIPTION
**Pull request by vodik**
There is no downside to this change:

- Lambdas and functions has identical performance characteristics on python 3 (and functions are slightly faster on python 2).
- It simplifies the compiler slightly.
- And there should be less surprises with the generated code (decorating lambas, for example, see #1506 ).

### CODE REVIEW

**Comment by Kodiologist** [comment]
If I'm not mistaken, this complicates the generated code: what could be a lambda will now take an extra assignment. Also, I'd like Hy to be able to generate every Python construct, even if it's not strictly necessary to do so (Python has a lot of constructs that are semantically redundant). Given, we don't have that yet (we're missing, e.g., variable annotations), but if we eliminate lambdas, we're taking a step back.

**Comment by tuturto** [comment]
But on the other hand, it simplifies code generation. In this case, that's a trade off I would be willing to make. If lambda is supported, we need to work around some special cases like decorating it (#1506).

**Comment by Kodiologist** [comment]
But it doesn't make sense to decorate a `fn`. You can just use a regular function call for that. You can decorate a `defn`, but that's already guaranteed not to produce a lambda.

**Comment by vodik** [comment]
But I think its weird that `fn` then *may* generate a function, if it detects it contains statements. Its certainly deterministic, but its an internal that leaks outwards.

As far as I'm concerned, conceptually, `fn` generates a function but may choose to pick the more "optimal" representation of a lambda. But when you can't use a lambda everywhere you'd use a function, small weird things arise.

Anyways, there is another approach to the other issue which I have a proof of concept: have a mechanism to promote `fn` to `fn*`.

**Comment by Kodiologist** [comment]
> As far as I'm concerned, conceptually, `fn` generates a function but may choose to pick the more "optimal" representation of a lambda.

It may be more useful to think of it the other way round. `fn` produces an anonymous function and returns it. That's the reason we have `fn`. If you want to assign your function a name from the beginning, instead of passing it as an argument or putting it in a list or something, you should probably be using `defn`. It is an implementation detail that sometimes `fn` needs to produce a Python `def` statement because of Python's limitations on what can go in a `lambda`.

**Comment by vodik** [comment]
Okay, so this is going to be a little on the stream of consciousness side, I apologize in advance.

Right, let me be clear, I understand that is how its intended to be interpreted, and its how the docs present it as well.

> But it doesn't make sense to decorate a `fn`...

But there's also a lot of effort put into rewriting `setv` + `fn`. This example:

```clojure
(setv foo (with-decorator bar (fn [x] (import x) x)))
```

emits equivalent code to:

```python
@bar
def foo(x):
    import x
    return x
```

Treating Hy like a black box, when I seeing this much effort in emitting optimal AST, I can't help but start to think of it the other way: that lambda is the optimization.

If we're going to say `fn` shouldn't be decorated, maybe we shouldn't let it be decorated at all. But I honestly don't see anything wrong with allowing `fn` to be decorated:

- We can force a function expansion
- We can make `with-decorator` work with function calls instead of decorator syntax when its a lambda (maybe this is the best solution `#@(classmethod my-decorator foobar) -> classmethod(my_decorator(lambda ...))`

> Python has a lot of constructs that are semantically redundant

My understanding is that that they're literally the same thing since Python 3 - internally at least.

> Also, I'd like Hy to be able to generate every Python construct, even if it's not strictly necessary to do so

Well, what about if we promoted `#%` and supported it first party to generate lambda's instead like in Clojure?

Anyways, its no hill I want to die on, what do you guys think about my ideas on augmenting `with-decorator` then?

**Comment by gilch** [comment]
I'm thinking of putting decorators inside the `defn` form itself, instead of wrapping them in another special form. #952. But `fn` wouldn't have this.

**Comment by Kodiologist** [comment]
> But there's also a lot of effort put into rewriting `setv` + `fn`

This might have been a mistake. I mean, it would be more intuitive if `defn` were a special form, replacing `fn*`, rather than `fn*` and `defn` both existing just to guarantee the creation of a Python `def`. I'm not sure how we got here. Full disclosure: I created `fn*`, but that was as part of a change from an earlier even less sensible design, which included `lambda` as a separate special form (#1228; see also the follow-up #1259).

> Well, what about if we promoted `#%` and supported it first party to generate lambda's instead like in Clojure?

I think that would just pass the buck, in a sense: like `fn` does now, `#%` would be able to produce a lambda in some cases and forced to produce in a `def` in others.

> Anyways, its no hill I want to die on, what do you guys think about my ideas on augmenting `with-decorator` then?

You can see how in #1228 I generalized `with-decorator` only to revert it in #1259. Since it's only necessary for `def` blocks, it makes sense to only support `def` blocks.

In conclusion, it looks like what got us down this path was the fact that `(#@(foo (fn …)))` works inconsistently. The easiest solution to that would be to make `#@` or `with-decorator` raise an error saying that you can just write `(foo (fn …))`.
